runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,180 @@
+local M = {}
+local util = require "vim.lsp.util"
+local highlight = require "vim.highlight"
+local ns = vim.api.nvim_create_namespace("lsp-semantic-tokens")
+
+local semantic_tokens = {}
+local last_tick = {}
+
+local function modifiers_from_number(x, modifiers_table)
+  local function get_bit(n, k)
+    -- Based on/from https://stackoverflow.com/a/26230537
+    -- (n >> k) & 1
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  end
+
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+M.token_map = {
+  type = "LspType",
+  class = "LspClass",
+  enum = "LspEnum",
+  interface = "LspInterface",
+  struct = "LspStruct",
+  typeParameter = "LspTypeParameter",
+  parameter = "LspParameter",
+  variable = "LspVariable",
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+---@private
+local function make_cache(base_map)
+  return setmetatable({}, { __index = function(outer_table, ft)
+    local rtn = setmetatable({}, { __index = function(inner_table, key)
+      local hl = base_map[key]
+      if hl then
+        hl = highlight._link_default_highlight_once(ft..hl, hl)
+        rawset(inner_table, key, hl)
+        return hl
+      end
+    end})
+    rawset(outer_table, ft, rtn)
+    return rtn
+  end})
+end
+
+local token_cache = make_cache(M.token_map)
+local modifiers_cache = make_cache(M.modifiers_map)
+
+---@private
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
+end
+
+---@private
+local function highlight_token(buf, token, token_cache, modifiers_cache)
+  local hl = token_cache[token.type]
+  if hl then
+    highlight(buf, token, hl)
+  end
+  for _, m in pairs(token.modifiers) do
+    hl = modifiers_cache[m]
+    if hl then
+      highlight(buf, token, hl)
+    end
+  end
+end
+
+function M._handle_full(response, ctx, _)
+  vim.api.nvim_buf_clear_namespace(ctx.bufnr, ns, 0, -1)
+  -- if tick is has changed our response is outdated!
+  if last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
```

Comment:
How to have a last_tick per client? Can we always be sure that the request arrive in the order that they have been sent to the LSP?

Reply:
I should compare the response id and save the request id or prevent that a token request is made while a response is still awaited.

Probably, a changed tick can probably resolved by a delta request.

Reply:
there seems to be a bug in the current changed tick method implementation

Reply:
@tjdevries @clason I need some help here. How can I check whether the text has changed since the moment I made the request? Is there any possibility to make sure the document version of the language server when I made the request is the same as the current state of the buffer? I could just `pcall` the highlights but that would be a hack.

Reply:
I think the same problem applies to `apply_text_edit`, does it not? I don't think we handle it there, where there is actual danger to the buffer (instead of intermittent weird highlights).

Reply:
I''ll have a look how document version is handled right now.

Reply:
The current solution (especially with the recently merged cjson) did cause any errors when I used this branch during work last week. It is certainly not ideal, but as @clason mentioned should this probably solved on a more fundamental level, e.g. by an opt-in possibility on request creation to the automatically map responses to errors if they are out-dated.